File: Leetcode/Array&String/49_group-anagrams/main.cpp
// 49. 字母异位词分组
// Created by bob on 2022/4/22.
//
#include<vector>
#include<string>
#include<unordered_map>

using namespace std;

class Solution {
public:
vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> mp;
    for (string &str: strs) {
        string key = str;
        sort(key.begin(), key.end());
        mp[key].emplace_back(str);
    }
    vector<vector<string>> ans;
    for (auto it=mp.begin(); it!=mp.end();it ++) {
        ans.emplace_back(it->second);
    }
    return ans;
}
};

int main() {
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<vector<string>> res = Solution().groupAnagrams(strs);    // expect: [["bat"],["nat","tan"],["ate","eat","tea"]]
    return 0;
}

File: Leetcode/Array&String/49_group-anagrams/main.py
import collections
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        mp = collections.defaultdict(list)
        for st in strs:
            counts = [0] * 26
            for ch in st:
                counts[ord(ch) - ord("a")] += 1
            # 需要将 list 转换成 tuple 才能进行哈希
            mp[tuple(counts)].append(st)

        return list(mp.values())





if __name__ == '__main__':
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    res = Solution().groupAnagrams(strs)
    print(res)